// @jup-ag/lend/dist/shared/lend.BzG5ldOV.mjs
import { PublicKey } from '@solana/web3.js';
import { a as address } from './lend.CioR9-te.mjs';

const LIQUIDITY_PROGRAM_ID = new PublicKey(address);
const getSupportedTokenList = () => {
  return PublicKey.findProgramAddressSync(
    [Buffer.from("supported_token_list")],
    LIQUIDITY_PROGRAM_ID
  )[0];
};
const getTokenReserveFromAsset = (asset) => {
  return PublicKey.findProgramAddressSync(
    [Buffer.from("reserve"), asset.toBuffer()],
    LIQUIDITY_PROGRAM_ID
  )[0];
};
const getUserSupplyPosition = (asset, protocol) => {
  return PublicKey.findProgramAddressSync(
    [
      Buffer.from("user_supply_position"),
      asset.toBuffer(),
      protocol.toBuffer()
    ],
    LIQUIDITY_PROGRAM_ID
  )[0];
};
const getUserBorrowPosition = (asset, protocol) => {
  return PublicKey.findProgramAddressSync(
    [
      Buffer.from("user_borrow_position"),
      asset.toBuffer(),
      protocol.toBuffer()
    ],
    LIQUIDITY_PROGRAM_ID
  )[0];
};
const getRateModel = (asset) => {
  return PublicKey.findProgramAddressSync(
    [Buffer.from("rate_model"), asset.toBuffer()],
    LIQUIDITY_PROGRAM_ID
  )[0];
};
const getReserve = (asset) => {
  return PublicKey.findProgramAddressSync(
    [Buffer.from("reserve"), asset.toBuffer()],
    LIQUIDITY_PROGRAM_ID
  )[0];
};
const getLiquidity = () => {
  return PublicKey.findProgramAddressSync(
    [Buffer.from("liquidity")],
    LIQUIDITY_PROGRAM_ID
  )[0];
};

const liquidity = {
  __proto__: null,
  getLiquidity: getLiquidity,
  getRateModel: getRateModel,
  getReserve: getReserve,
  getSupportedTokenList: getSupportedTokenList,
  getTokenReserveFromAsset: getTokenReserveFromAsset,
  getUserBorrowPosition: getUserBorrowPosition,
  getUserSupplyPosition: getUserSupplyPosition
};

export { getRateModel as a, getUserBorrowPosition as b, getUserSupplyPosition as c, getTokenReserveFromAsset as d, getReserve as e, getLiquidity as g, liquidity as l };
