// @jup-ag/lend/dist/refinance/index.mjs
import { v as vaultsIdl } from '../shared/lend.C2-jCLFw.mjs';
import { e as getOperateIx } from '../shared/lend.yP9-PjAh.mjs';
import { PublicKey } from '@solana/web3.js';
import { Program } from '@coral-xyz/anchor';
import '../shared/lend.CioR9-te.mjs';
import '@solana/spl-token';
import 'bn.js';
import '../shared/lend.BzG5ldOV.mjs';

const address = "jup8QcdtqecBGw1iXHW3hQAsHQbTgEqbLbNMvvULmeK";
const metadata = {
	name: "refinance",
	version: "0.1.0",
	spec: "0.1.0",
	description: "Created with Anchor"
};
const instructions = [
	{
		name: "operate",
		discriminator: [
			217,
			106,
			208,
			99,
			116,
			151,
			42,
			135
		],
		accounts: [
			{
				name: "signer",
				writable: true,
				signer: true
			},
			{
				name: "signer_supply_token_account",
				writable: true
			},
			{
				name: "signer_borrow_token_account",
				writable: true
			},
			{
				name: "recipient",
				optional: true
			},
			{
				name: "recipient_borrow_token_account",
				optional: true
			},
			{
				name: "recipient_supply_token_account",
				optional: true
			},
			{
				name: "vault_config",
				docs: [
					"@dev mut because this PDA signs the CPI to liquidity program",
					"@dev verification inside instruction logic"
				]
			},
			{
				name: "vault_state",
				docs: [
					"@dev verification inside instruction logic"
				],
				writable: true
			},
			{
				name: "supply_token"
			},
			{
				name: "borrow_token"
			},
			{
				name: "oracle"
			},
			{
				name: "position",
				writable: true
			},
			{
				name: "position_token_account",
				docs: [
					"@dev verification inside instruction logic"
				]
			},
			{
				name: "current_position_tick",
				writable: true
			},
			{
				name: "final_position_tick",
				writable: true
			},
			{
				name: "current_position_tick_id"
			},
			{
				name: "final_position_tick_id",
				writable: true
			},
			{
				name: "new_branch",
				writable: true
			},
			{
				name: "supply_token_reserves_liquidity",
				writable: true
			},
			{
				name: "borrow_token_reserves_liquidity",
				writable: true
			},
			{
				name: "vault_supply_position_on_liquidity",
				writable: true
			},
			{
				name: "vault_borrow_position_on_liquidity",
				writable: true
			},
			{
				name: "supply_rate_model"
			},
			{
				name: "borrow_rate_model"
			},
			{
				name: "vault_supply_token_account",
				writable: true
			},
			{
				name: "vault_borrow_token_account",
				writable: true
			},
			{
				name: "supply_token_claim_account",
				writable: true,
				optional: true
			},
			{
				name: "borrow_token_claim_account",
				writable: true,
				optional: true
			},
			{
				name: "liquidity"
			},
			{
				name: "liquidity_program"
			},
			{
				name: "oracle_program"
			},
			{
				name: "supply_token_program"
			},
			{
				name: "borrow_token_program"
			},
			{
				name: "associated_token_program"
			},
			{
				name: "system_program",
				address: "11111111111111111111111111111111"
			},
			{
				name: "vaults_program",
				address: "jupr81YtYssSyPt8jbnGuiWon5f6x9TcDEFxYe3Bdzi"
			}
		],
		args: [
			{
				name: "new_col",
				type: "i128"
			},
			{
				name: "new_debt",
				type: "i128"
			},
			{
				name: "transfer_type",
				type: {
					option: {
						defined: {
							name: "TransferType"
						}
					}
				}
			},
			{
				name: "remaining_accounts_indices",
				type: "bytes"
			}
		]
	}
];
const types = [
	{
		name: "TransferType",
		type: {
			kind: "enum",
			variants: [
				{
					name: "SKIP"
				},
				{
					name: "DIRECT"
				},
				{
					name: "CLAIM"
				}
			]
		}
	}
];
const refinanceIdl = {
	address: address,
	metadata: metadata,
	instructions: instructions,
	types: types
};

const getRefinanceProgram = ({
  connection,
  signer
}) => new Program(refinanceIdl, {
  connection,
  publicKey: signer
});
async function getRefinanceIx(params) {
  let {
    nftId,
    accounts,
    remainingAccounts,
    remainingAccountsIndices,
    ixs,
    addressLookupTableAccounts,
    addressLookupTableAddresses
  } = await getOperateIx(params);
  accounts = {
    ...accounts,
    vaultsProgram: new PublicKey(vaultsIdl.address)
  };
  const program = getRefinanceProgram({
    connection: params.connection,
    signer: params.signer
  });
  const operateIx = await program.methods.operate(
    params.colAmount,
    params.debtAmount,
    { direct: {} },
    Buffer.from(remainingAccountsIndices)
  ).accounts(accounts).remainingAccounts(remainingAccounts).instruction();
  return {
    nftId,
    accounts,
    remainingAccounts,
    remainingAccountsIndices,
    addressLookupTableAccounts,
    addressLookupTableAddresses,
    otherIxs: ixs.slice(0, -1),
    operateIx
  };
}

export { getRefinanceIx };
