// @jup-ag/lend/dist/api/index.mjs
import axios from 'axios';

class BorrowClient {
  http;
  constructor({ http }) {
    this.http = http;
  }
  async getVaults() {
    const { data } = await this.http.get("/v1/borrow/vaults");
    return data;
  }
  async operate({
    signer,
    colAmount,
    debtAmount,
    vaultId,
    positionId,
    positionOwner
  }) {
    const { data } = await this.http.post("/v1/borrow/operate", {
      signer,
      colAmount,
      debtAmount,
      vaultId,
      positionId,
      positionOwner
    });
    return data;
  }
  async operateInstructions({
    signer,
    colAmount,
    debtAmount,
    vaultId,
    positionId,
    positionOwner
  }) {
    const { data } = await this.http.post(
      "/v1/borrow/operate-instructions",
      {
        signer,
        colAmount,
        debtAmount,
        vaultId,
        positionId,
        positionOwner
      }
    );
    return data;
  }
}

class EarnClient {
  http;
  constructor({ http }) {
    this.http = http;
  }
  async getTokens() {
    const { data } = await this.http.get("/v1/earn/tokens");
    return data;
  }
  async getEarnings({
    user,
    positions
  }) {
    const { data } = await this.http.get(
      "/v1/earn/earnings",
      {
        params: {
          user,
          positions
        }
      }
    );
    return data;
  }
  async getPositions({ users }) {
    const { data } = await this.http.get(
      "/v1/earn/positions",
      {
        params: {
          users
        }
      }
    );
    return data;
  }
  async deposit({
    asset,
    signer,
    amount
  }) {
    const { data } = await this.http.post("/v1/earn/deposit", {
      asset,
      signer,
      amount
    });
    return data;
  }
  async depositInstructions({
    asset,
    signer,
    amount
  }) {
    const { data } = await this.http.post(
      "/v1/earn/deposit-instructions",
      {
        asset,
        signer,
        amount
      }
    );
    return data;
  }
  async withdraw({
    asset,
    signer,
    amount
  }) {
    const { data } = await this.http.post("/v1/earn/withdraw", {
      asset,
      signer,
      amount
    });
    return data;
  }
  async withdrawInstructions({
    asset,
    signer,
    amount
  }) {
    const { data } = await this.http.post(
      "/v1/earn/withdraw-instructions",
      {
        asset,
        signer,
        amount
      }
    );
    return data;
  }
  async mint({
    asset,
    signer,
    shares
  }) {
    const { data } = await this.http.post("/v1/earn/mint", {
      asset,
      signer,
      shares
    });
    return data;
  }
  async mintInstructions({
    asset,
    signer,
    shares
  }) {
    const { data } = await this.http.post(
      "/v1/earn/mint-instructions",
      {
        asset,
        signer,
        shares
      }
    );
    return data;
  }
  async redeem({
    asset,
    signer,
    shares
  }) {
    const { data } = await this.http.post("/v1/earn/redeem", {
      asset,
      signer,
      shares
    });
    return data;
  }
  async redeemInstructions({
    asset,
    signer,
    shares
  }) {
    const { data } = await this.http.post(
      "/v1/earn/redeem-instructions",
      {
        asset,
        signer,
        shares
      }
    );
    return data;
  }
}

class Client {
  borrowClient;
  earnClient;
  http;
  constructor({ apiKey } = {}) {
    this.http = axios.create({
      baseURL: apiKey ? `https://api.jup.ag/lend` : "https://lite-api.jup.ag/lend",
      headers: {
        "x-api-key": apiKey || ""
      }
    });
    this.borrowClient = new BorrowClient({ http: this.http });
    this.earnClient = new EarnClient({ http: this.http });
  }
  get borrow() {
    return this.borrowClient;
  }
  get earn() {
    return this.earnClient;
  }
}

export { Client };
