// @jup-ag/lend/dist/merkle-distributor/index.mjs
import { SystemProgram } from '@solana/web3.js';
import { g as getClaimStatus, a as getMerkleAdmin, b as getRewardsDistributor, c as merkleDistributor } from '../shared/lend.Bpqe1Iia.mjs';
import { Program } from '@coral-xyz/anchor';
import { ASSOCIATED_TOKEN_PROGRAM_ID, getAssociatedTokenAddressSync } from '@solana/spl-token';
import 'bn.js';

const getAccountOwner = async (account, connection) => {
  const info = await connection.getAccountInfo(account);
  if (!info)
    throw new Error(`Account info not found for ${account.toString()}`);
  return info.owner;
};
async function getClaimRewardsContext({
  rewardToken,
  distributorId,
  positionId,
  connection,
  signer
}) {
  const tokenProgram = await getAccountOwner(rewardToken, connection);
  return {
    signer,
    merkleDistributor: getRewardsDistributor(
      rewardToken,
      distributorId
    ),
    merkleAdmin: getMerkleAdmin(),
    mint: rewardToken,
    claimStatus: getClaimStatus(signer, positionId),
    vaultTokenAccount: getAssociatedTokenAddressSync(
      rewardToken,
      getMerkleAdmin(),
      true,
      tokenProgram
    ),
    recipientTokenAccount: getAssociatedTokenAddressSync(
      rewardToken,
      signer,
      false,
      tokenProgram
    ),
    tokenProgram,
    associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
    systemProgram: SystemProgram.programId
  };
}
async function getClaimIx({
  cumulativeAmount,
  positionType,
  positionId,
  cycle,
  merkleProof,
  metadata,
  signer,
  rewardToken,
  distributorId,
  connection
}) {
  const program = new Program(merkleDistributor, {
    connection,
    publicKey: signer
  });
  const accounts = await getClaimRewardsContext({
    rewardToken,
    distributorId,
    positionId,
    connection,
    signer
  });
  const ix = await program.methods.claim(
    cumulativeAmount,
    positionType,
    positionId,
    cycle,
    merkleProof.map((proof) => [...proof]),
    metadata
  ).accounts(accounts).instruction();
  return ix;
}

export { getClaimIx, getClaimRewardsContext };
