// @jup-ag/lend/dist/flashloan/index.mjs
import { SYSVAR_INSTRUCTIONS_PUBKEY, SystemProgram, PublicKey } from '@solana/web3.js';
import BN from 'bn.js';
import { ASSOCIATED_TOKEN_PROGRAM_ID, getAssociatedTokenAddressSync } from '@solana/spl-token';
import { g as getLiquidity, a as getRateModel, b as getFlashloanBorrowPosition, c as getLiquidityReserve, d as getFlashloanAdmin, e as flashloan } from '../shared/lend.Cr2l14_0.mjs';
import { l as liquidity } from '../shared/lend.CioR9-te.mjs';
import { Program } from '@coral-xyz/anchor';

const LIQUIDITY_PROGRAM_ID = new PublicKey(liquidity.address);
const getAccountOwner = async (account, connection) => {
  const info = await connection.getAccountInfo(account);
  if (!info)
    throw new Error(`Account info not found for ${account.toString()}`);
  return info.owner;
};
async function getFlashloanContext({
  signer,
  asset,
  connection
}) {
  const tokenProgram = await getAccountOwner(asset, connection);
  const liquidityKey = getLiquidity();
  return {
    signer,
    flashloanAdmin: getFlashloanAdmin(),
    signerBorrowTokenAccount: getAssociatedTokenAddressSync(
      asset,
      signer,
      false,
      tokenProgram
    ),
    mint: asset,
    flashloanTokenReservesLiquidity: getLiquidityReserve(asset),
    flashloanBorrowPositionOnLiquidity: getFlashloanBorrowPosition(asset),
    rateModel: getRateModel(asset),
    vault: getAssociatedTokenAddressSync(
      asset,
      liquidityKey,
      true,
      tokenProgram
    ),
    liquidity: getLiquidity(),
    liquidityProgram: new PublicKey(LIQUIDITY_PROGRAM_ID),
    tokenProgram,
    associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
    systemProgram: SystemProgram.programId,
    instructionSysvar: SYSVAR_INSTRUCTIONS_PUBKEY
  };
}
async function getFlashBorrowIx(params) {
  const { connection, signer } = params;
  const program = new Program(flashloan, {
    connection,
    publicKey: signer
  });
  const context = await getFlashloanContext(params);
  return await program.methods.flashloanBorrow(params.amount).accounts(context).instruction();
}
async function getFlashPaybackIx(params) {
  const { connection, signer } = params;
  const program = new Program(flashloan, {
    connection,
    publicKey: signer
  });
  const context = await getFlashloanContext(params);
  return await program.methods.flashloanPayback(params.amount).accounts(context).instruction();
}
async function getFlashloanIx(params) {
  const { connection, signer } = params;
  const program = new Program(flashloan, {
    connection,
    publicKey: signer
  });
  const [flashloanAdmin, borrowIx] = await Promise.all([
    program.account.flashloanAdmin.fetch(getFlashloanAdmin()),
    getFlashBorrowIx(params)
  ]);
  const paybackAmount = params.amount.mul(
    new BN(1e4 + flashloanAdmin.flashloanFee).div(new BN(1e4))
  );
  const paybackIx = await getFlashPaybackIx({
    ...params,
    amount: paybackAmount
  });
  return { borrowIx, paybackIx };
}

export { getFlashBorrowIx, getFlashPaybackIx, getFlashloanContext, getFlashloanIx };
